---
title: "Using R & Bioconductor to assign cell types to single-cell RNA-seq data"
output: 
  html_document:
    toc: true
    toc_float: true
author: Niyati Saini
date: " 06 March, 2025"

---

# Code adapted from: http://kieranrcampbell.github.io/r-workshop-march-2019
# Modifications: Added updated scater functions, replaced deprecated methods

```{r}
setwd("~/Downloads/r-workshop-march-2019-master")
getwd() # Confirm the correct directory
```

## Required packages

```{r}
#install.packages("BiocManager")
#BiocManager::install("DropletUtils")

suppressPackageStartupMessages({
  library(scater) # BioConductor
  library(SingleCellExperiment) # BioConductor
  library(DropletUtils) # BioConductor
  library(tidyverse) # CRAN
  library(here) # CRAN
  library(DT) # CRAN
  library(pheatmap) # CRAN
})

knitr::opts_chunk$set(echo = TRUE,
                      cache = TRUE)
```

```{r}
#install.packages(c("tidyverse", "here", "DT", "pheatmap", "BiocManager"))
#BiocManager::install(c("scater", "SingleCellExperiment", "DropletUtils"))
#install.packages("tensorflow", repos = "http://cran.us.r-project.org")
#tensorflow::install_tensorflow()
#install.packages("devtools", repos = "http://cran.us.r-project.org") # If not already installed
#devtools::install_github("Irrationone/cellassign")
```

# Read in 10X scRNA-seq data
```{r}
data_dir <- "/Users/niyati/Downloads/r-workshop-march-2019-master/data/outs/filtered_gene_bc_matrices/GRCh38"
print(data_dir)
print(dir(data_dir))
```

```{r}
sce <- read10xCounts(data_dir)
sce
```

# Getting data dimensions
```{r}
nrow(sce) # Number of rows = genes
ncol(sce) # Number of columns = cells
```

```{r}
print(sce)
```
## Subsetting
```{r}
sce[, c(1,3,5)] # Subset to cells 1, 3, 5
```

```{r}
sce[c(2,4,6), ] # Subset to genes 2, 4, 6 
```
## Feature and cell metadata
`rownames` = gene names
```{r}
head(rownames(sce))
```

`colnames` = cell names = barcodes (sometimes)
```{r}
head(colnames(sce))
```
#Column data (cell specific metadata)
```{r}
head(colData(sce))
```
# set the column names to the barcode:

```{r}
colnames(sce) <- colData(sce)$Barcode

head(colnames(sce))
```

# subset on barcode:
```{r}
sce[, "AAACCTGCAGTAAGCG-1"]
```

# Row data (gene specific metadata)
```{r}
head(rowData(sce))
```
# `reducedDims` is where PCA,UMAP,tSNE representations will live - but I haven't made them yet
```{r}
reducedDims(sce)
```

# `sizeFactors` is where cell size factors will live - but I haven't calculated them yet
```{r}
head(sizeFactors(sce))
```

# The ability to have multiple `assays` is one of the unique advantages of the SingleCellExperiment approach - I can carry around `counts`, `logcounts`, and any other weird data transformation I might like. Right now I only have raw counts:

```{r}
names(assays(sce))
```

```{r, eval = FALSE}
assay(sce, "counts")
counts(sce)
```
# I can make my own:

```{r}
assay(sce, "my_super_strange_assay") <- sin(as.matrix(counts(sce)))

names(assays(sce))
```

```{r}
class(assay(sce, "my_super_strange_assay"))
```

# Note the beauty of SingleCellExperiments is that subsetting is consistent: if I want to subset only some cells and genes:

```{r}
sce_subset <- sce[c(1,3,5), c(2,4,6,8)]
```

# Then everything else is subset too!

```{r}
print(dim(counts(sce_subset)))
print(length(sizeFactors(sce_subset)))
print(dim(rowData(sce_subset)))
```


